A72-open-telemetry-tracing.md

```diff
@@ -0,0 +1,680 @@
+from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator
+
+class OpenTelemetryPlugin:
+    """
+    If `tracer_provider` is None, no traces are collected.
+    """
+
+    def __init__(
+        self,
+        *,
+        tracer_provider: Optional[TracerProvider] = None,
+        text_map_propagator: [TraceContextTextMapPropagator] = None,
+    ):
+```
+
+### Go
+The following methods will be added to the opentelemetry package.
+```go
+import (
+  "go.opentelemetry.io/otel/trace"
+)
+
+// TraceOptions are the trace options for OpenTelemetry instrumentation.
+type TraceOptions struct {
+  TraceProvider trace.TraceProvider
+}
+
+// DialOption returns a dial option which enables OpenTelemetry instrumentation
+// code for a grpc.ClientConn.
+//
+// Client applications interested in instrumenting their grpc.ClientConn should
+// pass the dial option returned from this function as a dial option to
+// grpc.Dial().
+//
+// Using this option will always lead to instrumentation, however in order to
+// use the data a SpanExporter must be registered with the TraceProvider in the
+// TraceOption. Client side has retries, so a Unary and Streaming Interceptor are
+// registered to handle per RPC traces, and a Stats Handler is registered to handle
+// per RPC attempt trace. These three components registered work together in
+// conjunction, and do not work standalone.
+func DialOption(to TraceOptions) grpc.DialOption {}
+
+// ServerOption returns a server option which enables OpenTelemetry
+// instrumentation code for a grpc.Server.
+//
+// Server applications interested in instrumenting their grpc.Server should pass
+// the server option returned from this function as an argument to
+// grpc.NewServer().
+//
+// Using this option will always lead to instrumentation, however in order to
+// use the data a SpanExporter must be registered with the TraceProvider option.
+// Server side does not have retries, so a registered Stats Handler is the only
+// option that is returned.
+func ServerOption(to TraceOptions) grpc.ServerOption {}
+```
+
+## Tracing Information
+RPCs on the client side may undergo retry attempts, whereas on the server side, 
+they do not. gRPC records both per-call tracing details (on the parent span) 
+and per-attempt tracing details (on the attempt span) on the client side. 
+On the server side, there is only per-call traces.  With the new OpenTelemetry 
+plugin we will produce the following tracing information during an RPC lifecycle:
+
+At the client, on parent span:
+* If the RPC experienced name resolution delay, add an Event at the start of the
+  call with the name "Delayed name resolution complete" upon completion of the
+  name resolution process.
+* When the call is closed, set RPC status and end the parent span. gRPC status "OK" 
+  is recorded with status "OK", while other gRPC statuses are marked as "ERROR". 
+  Non-"OK" statuses include their code as a description, the same below.
```

Comment:
Are we not including the status description? Seems that is useful, and was included with OpenCensus.

Reply:
oh this may have to change if we don't want to overload description with both description and status code.
What about inherit `opentelemetry.StatuCode` and add grpc statuscode there?

opentelemetry does not have API to specify detailed status code(their enum only has UNSET OK ERROR)
https://javadoc.io/doc/io.opentelemetry/opentelemetry-api/latest/io/opentelemetry/api/trace/Span.html#setStatus(io.opentelemetry.api.trace.StatusCode,java.lang.String)

Reply:
I gave an example of overloading the description with grpc status code.